### ISSUE TITLE
After pdm version 2.13+ `pdm update` raises `ValueError: unicode string is too large` with some packages

### ISSUE THREAD

**Issue by wikiped** [pdm-project/pdm#2774]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

## Steps to reproduce

1. Make sure `pdm -V` shows version above 2.13: `PDM, version 2.13.2`

2. Execute commands:
  `pdm init -p issue_2774 --lib`
  `cd ./issue_2774`
  `pdm add ruamel-yaml msgspec dateparser`

## Actual behavior

Error is thrown:

`ValueError: unicode string is too large`

See attached `pdm-lock.log` for complete log

Downgrading to pdm `2.12.4` (the last before 2.13 ?) solves the issue.

Difficult to say if there is anything specific to these (set of ?) packages of there might be other combinations causing the same error, but after updating pdm to `2.13` this error appeared in a project with a bunch of dependencies, which I have distilled to this minimal set of packages.

As a side note, perhaps worth of a separate issue, resolving these 3 (!!!) packages took a few minutes to fail. Imaging a dozen or more packages - it takes tens of minutes. Perhaps there is some room for improvement there...?

## Expected behavior

🎉 All complete!

## Environment Information

```powershell
# Paste the output of `pdm info && pdm info --env` below:
pdm info
PDM version:
  2.13.2
Python Interpreter:
  ...\issue_2768\.venv\Scripts\python.exe (3.12)
Project Root:
  .../issue_2768
Local Packages:

pdm info --env
{
  "implementation_name": "cpython",
  "implementation_version": "3.12.2",
  "os_name": "nt",
  "platform_machine": "AMD64",
  "platform_release": "10",
  "platform_system": "Windows",
  "platform_version": "10.0.19045",
  "python_full_version": "3.12.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.12",
  "sys_platform": "win32"
}

```
[pdm-lock-_w6_1itx.log](https://github.com/pdm-project/pdm/files/14886028/pdm-lock-_w6_1itx.log)

**Comment by frostming** [comment]
I am not able to reproduce, even enabling` pypi.json_api` as the error log shows.

It also shows that the error happens when PDM tries to serialize the response of a JSON API request. The body reaches the max string length limit 2^32-1 defined by msgpack, that is equivalent to 4 GiB. I've checked several packages and didn't find anything abnormal. The URL format is `https://pypi.org/pypi/<package>/<version>/json`.

So I can't get much further here. You can dig into what request caused the exception yourself, and what the response body was.

**Comment by wikiped** [comment]
Turning off `pypi.json_api` did help, thank you.

Although it does seem strange that previous version did not have any problems with that...

**Comment by elbaro** [comment]
I hit the same problem with `pypi.json_api` off. There is no way the existing dependency suddenly got over 2^32 (4GB).

Fixed by upgrading to 2.13.3 and `pdm cache clear`.

**Comment by frostming** [comment]
> it looked like a httpx Request object was self-referencing in the URL field, causing the serialization for it to explode in the MessagePack serializer.

That's very enlightening, can you find a way how to reproduce this kind of request?
